### ISSUE TITLE
Type Inference is unable to deduce indirect base class

### ISSUE THREAD

**Issue by JohannesRudolph** [Microsoft/TypeScript#13386]
**TypeScript Version:**  2.1.1 / nightly (2.2.0-dev.201xxxxx)
2.14

**Code**

```ts
interface HalObjectLinks {
    self: HalLink;
}

interface HalLink {
    href: string;
}

interface HalObject<R extends HalObjectLinks> {
    '_links': HalObjectLinks;
}

interface CustomerLinks extends HalObjectLinks {
    addresses: HalLink;
}

interface Customer extends HalObject<CustomerLinks> {
    name: string;
}

class CustomerModel implements Customer {
    public name: string;

    public '_links': HalObjectLinks = {
        self: { href: 'self' }
    };
}

function findHref<T extends HalObjectLinks, K extends keyof T>(object: HalObject<T>, rel: K): string {
    let link = object._links[<string>rel];
    if (!link) {
        throw new Error('HAL object did not have link: ' + rel);
    }

    return link.href;
}

function infer<T extends HalObjectLinks>(object: HalObject<T>): HalObject<T> {
    return object;
}

let a = new CustomerModel();
findHref(a, 'addresses'); // inference fails: 'Argument of type '"addresses"' is not assignable to parameter of type '"self"'.'


let x: HalObject<CustomerLinks> = a;
findHref(x, 'addresses'); // compiles fine

```

**Expected behavior:**
TSC should not flag an error and correctly infer that a is a `HalObject<CustomerLinks>` (indirect base classes above).

**Actual behavior:**
See comments in code

**Comment by HerringtonDarkholme** [comment]
Because

```ts
interface HalObject<R extends HalObjectLinks> {
    '_links': HalObjectLinks;
}
```

is not correct. Please see the correct version below, correction commented with `here` aside.

```ts
interface HalObjectLinks {
    self: HalLink;
}

interface HalLink {
    href: string;
}

interface HalObject<R extends HalObjectLinks> {
    '_links': R; // here
}

interface CustomerLinks extends HalObjectLinks {
    addresses: HalLink;
}

interface Customer extends HalObject<CustomerLinks> {
    name: string;
}

class CustomerModel implements Customer {
    public name: string;

    public '_links' = { // here
        self: { href: 'self' },
        addresses: {href: 'addr'} // here
    };
}

function findHref<T extends HalObjectLinks, K extends keyof T>(object: HalObject<T>, rel: K): string {
 // omitted 
  return ''
}

function infer<T extends HalObjectLinks>(object: HalObject<T>): HalObject<T> {
    return object;
}

let a = new CustomerModel();
findHref(a, 'addresses'); // inference fails: 'Argument of type '"addresses"' is not assignable to parameter of type '"self"'.'


let x: HalObject<CustomerLinks> = a;
findHref(x, 'addresses'); // compiles fine
```

**Comment by JohannesRudolph** [comment]
The beauty of the approach I did originally present was that it forces/nudges the consumer to use the `findHref` for everything except the self link, while preserving type information in the `findHref` call about the link properties which _should_ be there. I found it unintuitive that TS wasn't able to infer this, but then I think it's one of the consequences of TS being a structurally typed language.

**Comment by mhegazy** [comment]
Please see https://github.com/Microsoft/TypeScript/wiki/FAQ#why-doesnt-type-inference-work-on-this-interface-interface-foot---
